src/mmtk.rs

```diff
@@ -403,35 +401,22 @@ impl<VM: VMBinding> MMTK<VM> {
     }
 
     /// The application code has requested a collection. This is just a GC hint, and
-    /// we may ignore it.
+    /// we may ignore it (See `ignore_system_gc` and `full_heap_system_gc` in [`crate::util::options::Options`]).
     ///
     /// # Arguments
     /// * `tls`: The mutator thread that requests the GC
-    /// * `force`: The request cannot be ignored (except for NoGC)
-    /// * `exhaustive`: The requested GC should be exhaustive. This is also a hint.
-    pub fn handle_user_collection_request(
-        &self,
-        tls: VMMutatorThread,
-        force: bool,
-        exhaustive: bool,
-    ) {
+    pub fn handle_user_collection_request(&self, tls: VMMutatorThread) {
```

Comment:
Some VMs gives the user fine-grained control of whether to trigger a full-heap GC or a minor GC.  In CRuby, the [`GC::start`](https://docs.ruby-lang.org/en/3.3/GC.html#method-c-start) method uses the `full_mark` parameter for that.  The current master branch of `mmtk-ruby` ignores this because it still doesn't support StickyImmix, but will enable StickyImmix once we fix the side forwarding bits bug.  We should leave a public API for the VM to choose whether to perform an exhaustive GC.

Note: `memory_manager::handle_user_collection_request` is public, and `MMTK::handle_user_collection_request` is a public method, too.

The OpenJDK binding calls `memory_manager::handle_user_collection_request` which does not have the `force` and `exhaustive` parameters.  I suggest we leave `Option::full_heap_system_gc` and `MMTK::handle_user_collection_request` unchanged, but change `memory_manager::handle_user_collection_request` to pass `true` and `true` to `MMTK::handle_user_collection_request`.

Reply:
> change memory_manager::handle_user_collection_request to pass true and true to MMTK::handle_user_collection_request

In this case, MMTk basically ignores those two options: `ignore_system_gc` is ignored, as we always force GC, and `full_heap_system_gc` is also ignored as we always do full heap/exhaustive GC.

Reply:
No.  `MMTK::handle_user_collection_request` is public, too.  The Ruby VM, for example, can directly call `mmtk.handle_user_collection_request(tls, false, true)` to trigger full-heap GC, and call `mmtk.handle_user_collection_request(tls, false, false)` to trigger nursery GC.  OpenJDK can call `memory_manager::handle_user_collection_request(tls)` to use the default `force` and `exhaustive` arguments.

Reply:
Well, I think a more elegant solution is to change the argument of `MMTK::handle_user_collection_request` to take one single `UserCollectionArgs` value:

```rs
struct UserCollectionArgs {
    force: bool,
    exhaustive: bool,
}

impl default for UserCollectionArgs {
    fn default() -> Self {
        Self { force: true, exhaustive: true }
    }
}
```

This will allow the VM binding to always call `mmtk.handle_user_collection_request(tls, UserCollectionArgs::default())` if they want the default value, or call `mmtk.handle_user_collection_request(tls, UserCollectionArgs { exhaustive: true, ..Default::default() }` if they want to customise.  But that'll be an API-breaking change.